module_utils/network_lsr/ethtool.py

```diff
@@ -0,0 +1,48 @@
+# SPDX-License-Identifier: BSD-3-Clause
+
+import array
+import binascii
+import struct
+import fcntl
+import socket
+
+
+ETHTOOL_GPERMADDR = 0x00000020
+SIOCETHTOOL = 0x8946
+
+
+def get_perm_addr(ifname):
+    """
+    Return the Permanent address value for the specified interface using the
+    ETHTOOL_GPERMADDR ioctl command. The address is returned as string in
+    python2 so it is needed to convert it to a bytearray and convert it again
+    to a hex string again.
+
+    Please for further documentation, see:
+    https://github.com/torvalds/linux/blob/master/include/uapi/linux/ethtool.h#L734
+    https://github.com/torvalds/linux/blob/master/include/uapi/linux/ethtool.h#L1388
+    https://git.kernel.org/pub/scm/network/ethtool/ethtool.git/tree/ethtool.c#n4172
+    """
+    try:
+        sock = socket.socket(socket.AF_INET, socket.SOCK_STREAM)
+        sockfd = sock.fileno()
+        ifname = ifname.encode("utf-8")
```

Comment:
is utf-8 correct here? Can the interfaces contain other characters that ASCII?

Reply:
According to https://git.kernel.org/pub/scm/network/iproute2/iproute2.git/tree/lib/utils.c?id=1f420318bda3cc62156e89e1b56d60cc744b48ad#n827 it seems from the kernel is a char[] so interfaces cannot contain other characters that ASCII. I will do the proper encoding.

Reply:
the code you linked there shows the opposite.

Like most kernel names (interface names, filenames(!!)) these names are in no defined encoding at all... except, that they must be '\0' terminated and that certain ASCII characters are treated special (`'.'`, `'/'`). That means, you cannot use any arbitrary encoding (e.g. UTF-32 would blow up), but as long as you use an encoding that honours this minimal requirements (like some extension to ASCII (UTF-8, ISO/IEC 8859)), kernel doesn't care.

Of course, since somebody has to type these "names" into a keyboard and print them on the screen, having not defined encoding, is a serious problem. The general recommendation is of course to only use ASCII or valid UTF-8.

For that reason, NetworkManager exposed interface names as ANSI-C backslash encoded strings, that are really binary. E.g. on `nm_device_get_iface()`. Try `ip link add name $'d\xccf\\c' type dummy` and then look at `ip link` and `nmcli device` output.

On the other end, when you configure a NetworkManager profile, you can only use UTF-8 name. See https://gitlab.freedesktop.org/NetworkManager/NetworkManager/-/blob/01667694ca75cb3dff137a28313548a6cc4258e7/shared/nm-glib-aux/nm-shared-utils.c#L4748

Treating this as UTF-8 is IMO the best you can do. If the caller provides anything else, it's her problem.

Reply:
Nice, `ip link add dev 😎 type veth peer 🌞` works, too.

Reply:
Also interesting... by default `journalctl` will not print binary data. That means, if you create a non-UTF-8 interface name, NM will log that binary name, and journal will only say `[143B blob data]`. Pass `--all` to journalctl to see it. Of course, then the output of journalctl is no longer valid UTF-8.

Reply:
> Nice, `ip link add dev 😎 type veth peer 🌞` works, too.

I do not want a world where people name interfaces with emojis :-)